pkg/stanza/operator/transformer/recombine/recombine.go

```diff
@@ -267,47 +277,71 @@ func (r *Transformer) matchIndicatesLast() bool {
 
 // addToBatch adds the current entry to the current batch of entries that will be combined
 func (r *Transformer) addToBatch(_ context.Context, e *entry.Entry, source string) {
-	if _, ok := r.batchMap[source]; !ok {
-		r.batchMap[source] = []*entry.Entry{e}
+	if batch, ok := r.batchMap[source]; !ok {
+		r.batchMap[source] = &SourceBatch{
+			entries:                 []*entry.Entry{e},
+			recombineFieldTotalSize: 0,
+		}
 		if len(r.batchMap) >= r.maxSources {
 			r.Error("Batched source exceeds max source size. Flushing all batched logs. Consider increasing max_sources parameter")
 			r.flushUncombined(context.Background())
+			return
+		}
+	} else {
+		batch.entries = append(batch.entries, e)
+	}
+	sourceBatch := r.batchMap[source]
+
+	if r.maxLogSize > 0 {
+		var s string
+		err := e.Read(r.combineField, &s)
```

Comment:
It's unfortunate that we have to read each log twice. Once here, and once in `flushSource`. Additionally, there is some non-trivial amount of duplicated code as well.

What do you think about pulling the `strings.Builder` into `sourceBatch`? This should allow us to read each string one and immediately add it with `builder.WriteString`. It also exposes a [Len](https://pkg.go.dev/strings#Builder.Len) method so we do not need to maintain a separate value on our own. Then, in `flushSource`, we only need to call `String()` on the builder and overwrite the field in the appropriate entry (according to `overwriteWithOldest` setting).

Reply:
Thanks a lot!!  I've modified the code following your suggestion.